### ISSUE TITLE
Run fewer convolutions in training

### ISSUE THREAD

**Issue by stsievert** [stsievert/scipy-convolve-timings#3]
The goal of this repo is to estimate `np.sign(direct_time - fft_time)`.

One way to do this is to accurately estimate `direct_time` and `fft_time`. This will require running convolutions for every shape. This is inefficient: a lot of (potentially unnecessary) convolutions are run.

Another way to do is with "uncertainty sampling" and only run the convolutions where `np.abs(direct_time - fft_time)` is about 0. This might yield the same accuracy on the test data (which should always be complete) despite having lower training time.

> Now that I think about it again though, a better approach would be to focus on fitting the curve near to the cross-over point as that's the only place where precision is actually important.
>
> —@peterbell10 in https://github.com/scipy/scipy/pull/10904#issuecomment-539886619

**Comment by stsievert** [comment]
I think a better approach would be to label this as a classification problem, not a regression problem. I imagine having labels of ±1 if `fftconvolve` is faster/slower, and having a two dimensional feature with number of floating point operations.

I would implement this, but I'm seeing satisfactory performance with the regression formulation. I'll leave this as future work.

Specifically, here's the data I would create:

<details>

``` python
def _get_flops(shape, method):
    # approximately
    flop = {"direct": n**2, "fft": n*np.log(n)}
    return flop[method]

def _get_faster_method(shape):
    """ Time convolution with shape for different methods, and report the faster one """
    ...
    return "fft" if fft_time < direct_time else direct

shapes = np.array([(3, 256), (5, 1024), (9, 512)])  # kernel: 3x3, image: 256x256
df = pd.DataFrame({
    "fft_ops": [_get_flop(s, "fft") for s in shapes],
    "direct_ops": [_get_flop(s, "direct") for s in shapes]
})
df["fastest"] = [_get_faster_method(s) for s in shapes]
labels = {"fft": 1, "direct": -1}
df["label"] = df.fastest.apply(labels.get)
```

</details>

And here's the classifier I would use:

``` python
from sklearn.linear_model import LogisticRegression
est = LogisticRegression(penalty="none", fit_intercept=False)
print(df.columns)  # fft_ops, direct_ops, fastest, label
X = df[["fft_ops", "direct_ops"]]
y = df["label"]
est.fit(X, y)
```

Then, I'd use the two resulting coefficients as big-O constants and integrate that into the prediction with `"fft" if O_fft * fft_ops < O_direct * direct_ops else "direct"`.

**Comment by stsievert** [comment]
This issue is meaningless now; I fit big-O constants separately for `method='fft'` and `method='direct'`.
